src/alerter.rs

```diff
@@ -0,0 +1,389 @@
+use crate::WatchtowerConfig;
+use crate::pagerduty::PagerDutyClient;
+
+use anyhow::Result;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, Instant};
+use tokio::sync::mpsc::{self, UnboundedSender, UnboundedReceiver};
+use tokio::sync::Mutex;
+
+#[derive(Deserialize, Clone, PartialEq, Eq, Debug, Default)]
+pub enum AlertLevel {
+    #[default]
+    None,
+    Info,
+    Warn,
+    Error,
+}
+
+#[derive(Clone, Debug)]
+pub struct AlertParams {
+    name: String,
+    description: String,
+    alert_level: AlertLevel,
+}
+
+impl AlertParams {
+    pub fn new(name: String, description: String, alert_level: AlertLevel) -> Self {
+        AlertParams { name, description, alert_level }
+    }
+
+    #[cfg(test)]
+    pub fn is_name_equal(&self, name: &str) -> bool {
+        self.name == name
+    }
+
+    #[cfg(test)]
+    pub fn is_description_equal(&self, description: &str) -> bool {
+        self.description == description
+    }
+
+    #[cfg(test)]
+    pub fn is_level_equal(&self, alert_level: AlertLevel) -> bool {
+        self.alert_level == alert_level
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct WatchtowerAlerter{
+    alert_sender: UnboundedSender<AlertParams>,
+    alert_receiver: Arc<Mutex<UnboundedReceiver<AlertParams>>>,
+    alert_cache: Arc<Mutex<HashMap<String, Instant>>>,
+    alert_cache_expiry: Duration,
```

Comment:
Consider prefixing these with `pagerduty_...`

Reply:
they might now so I m leaving as is